jlm/util/HashSet.hpp

```diff
@@ -293,6 +293,35 @@ class HashSet
     other.Clear();
   }
 
+  /**
+   * Modifies this HashSet object by removing any elements that are present in \p other.
+   *
+   * @param other the HashSet used as the negative side of the set difference.
+   */
+  void
+  DifferenceWith(const HashSet<ItemType> & other)
+  {
+    // If this HashSet is smaller, loop over it and remove elements in other.
+    // If other is smaller, loop over it and remove elements from this.
+
+    if (Size() <= other.Size())
+    {
+      // This branch also handles the unlikely case where this and other are the same set.
+
+      auto inOther = [&](const ItemType & item)
+      {
+        return other.Contains(item);
+      };
+
+      RemoveWhere(inOther);
+    }
+    else
+    {
+      for (auto & item : other.Set_)
+        Remove(item);
```

Comment:
Could this not be: `Set_.erase(other.Set_.begin(), other.Set_.end())`. It might be a bit faster than iterating over all the elements individually.

Reply:
No, the documentation for `erase(first, last)` specifics that it must be iterators within the same set.

Reply:
Meh....